pkg/cloudprovider/provider/vultr/provider.go

```diff
@@ -168,52 +245,168 @@ func (p *provider) Validate(ctx context.Context, _ *zap.SugaredLogger, spec clus
+		return nil, vltErrorToTerminalError(resp.StatusCode, err)
 	}
 
 	for _, instance := range instances {
-		for _, tag := range instance.Tags {
-			if tag == string(machine.UID) {
-				return &vultrInstance{instance: &instance}, nil
-			}
+		if sets.NewString(instance.Tags...).Has(string(machine.UID)) &&
+			instance.Label == machine.Name {
+			return &vultrVirtualMachine{instance: &instance}, nil
 		}
 	}
 
 	return nil, cloudprovidererrors.ErrInstanceNotFound
 }
 
 func (p *provider) Get(ctx context.Context, _ *zap.SugaredLogger, machine *clusterv1alpha1.Machine, _ *cloudprovidertypes.ProviderData) (instance.Instance, error) {
-	return p.get(ctx, machine)
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+	if !c.PhysicalMachine {
+		return p.getVirtualMachine(ctx, c, machine)
+	}
+
+	return p.getPhysicalMachine(ctx, c, machine)
 }
 
 func (p *provider) GetCloudConfig(_ clusterv1alpha1.MachineSpec) (config string, name string, err error) {
 	return "", "", nil
 }
 
-func (p *provider) Create(ctx context.Context, _ *zap.SugaredLogger, machine *clusterv1alpha1.Machine, _ *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+func (p *provider) waitForInstanceCreation(ctx context.Context, client *govultr.Client, c *Config, instance instance.Instance, machine *clusterv1alpha1.Machine) error {
+	return wait.PollUntilContextTimeout(ctx, createCheckPeriod, createCheckTimeout, false, func(ctx context.Context) (bool, error) {
+		var err error
+		if !c.PhysicalMachine {
+			_, err = p.getVirtualMachine(ctx, c, machine)
+		} else {
+			_, err = p.getPhysicalMachine(ctx, c, machine)
+		}
+
+		if err != nil {
+			if isTerminalErr, _, _ := cloudprovidererrors.IsTerminalError(err); isTerminalErr {
+				return true, err
+			}
+			// Wait for some time as instance creation is successful
+			// just that we are not able to fetch it
+			time.Sleep(createCheckFailedWaitPeriod)
+			return false, fmt.Errorf("instance %q created but controller failed to fetch instance details", instance.Name())
+		}
+		return false, nil
+	})
+}
+
+func (p *provider) createVirtualMachine(ctx context.Context, client *govultr.Client, c *Config, machine *clusterv1alpha1.Machine, osid int, userdata string) (*vultrVirtualMachine, error) {
+
+	tags := sets.List[string](sets.New(c.Tags...).Insert(string(machine.UID)))
+
+	instanceCreateRequest := govultr.InstanceCreateReq{
+		Region: c.Region,
+		Plan:   c.Plan,
+		OsID:   osid,
+
+		Label:    machine.Spec.Name,
+		UserData: base64.StdEncoding.EncodeToString([]byte(userdata)),
+		Tags:     tags,
+
+		EnableIPv6: &c.EnableIPv6,
+		EnableVPC:  &c.EnableVPC,
+		AttachVPC:  c.VpcID,
+		EnableVPC2: &c.EnableVPC2,
+		AttachVPC2: c.Vpc2Id,
+	}
+	instance, resp, err := client.Instance.Create(ctx, &instanceCreateRequest)
```

Comment:
I might be unfamiliar with vultr API, so it might be a completely silly question, however, shouldn't be response body closed here, as I noticed you do that in some other cases?

Reply:
You are right, it should be closed. They are not closing it in the underlying function

Reply:
Fixed it.